backend/src/MotorDriverCardFactory.cc

```diff
@@ -24,28 +24,31 @@ namespace mtca4u {
       std::string mapModuleName,
       std::string motorConfigFileName) {
     boost::lock_guard<boost::mutex> guard(_factoryMutex);
-    // if the card is not in the map it will be created, otherwise the
-    // existing one is delivered
-    // TSK - fix me - now key must be a pair of deviceFileName and mapModuleName,
-    // not only deviceFileName
-    boost::shared_ptr<MotorDriverCard>& motorDriverCard = _motorDriverCards[std::make_pair(alias, mapModuleName)];
-
-    // create a new instance of the motor driver card if the shared pointer is
-    // empty, which means it has just been added to the map
-    if(!motorDriverCard) {
-      if(_dummyMode) { // if (deviceFileName=="custom://MotorDriverCardDummy")
-        // just create a MotorDriverCardDummy
-        motorDriverCard.reset(new MotorDriverCardDummy());
-      }
-      else {
-        boost::shared_ptr<ChimeraTK::Device> device(new ChimeraTK::Device);
-        device->open(alias);
-        MotorDriverCardConfig cardConfig = MotorDriverCardConfigXML::read(motorConfigFileName);
-
-        motorDriverCard.reset(new MotorDriverCardImpl(device, mapModuleName, cardConfig));
-      } // else deviceFileName=="/dummy/MotorDriverCard"
-
-    } // if (!motorDriverCard)
+
+
+    // Check if we have the card and if the weak reference is still valid. That way we can guarantee that there is only
+    // one MotorDriverCard instance accessing the device (at least in this process) while also being able to re-open
+    // the underlying device by giving up all motors using the card.
+    auto id = std::make_pair(alias, mapModuleName);
+    auto iterator = _motorDriverCards.find(id);
+    if(iterator != _motorDriverCards.end()) {
+      auto card = iterator->second.lock();
+      if(card) return card;
+    }
+
+    boost::shared_ptr<MotorDriverCard> motorDriverCard;
+    if(_dummyMode) {
+      // just create a MotorDriverCardDummy
+      motorDriverCard.reset(new MotorDriverCardDummy());
```

Comment:
I am a bit rusty on this: I think the new is not exception-safe, so one should use make_shared(). (?)

Reply:
According to https://stackoverflow.com/questions/20053504/stdshared-ptr-exception-safety this should properly delete the pointer if an exception occurs - at least with std::shared_ptr. Also make_shared does not work because the shared pointer is on the base class

Reply:
We can leave it like this. In this situation it is OK because we only have one objects that's being created.
The base class issue could have been worked around with a pointer cast:
```motorDriverCard = dynamic_pointer_cast< MotorDriverCard > (make_shared<MotorDriverCardDummy>());```
I am not even sure if the assignment of a shared_ptr of a base class is not working automatically.